From fc480b6dd60b810607e5d7e97b063b422dd3f671 Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Fri, 1 Sep 2017 13:30:29 +0800
Subject: [PATCH 2474/5242] MLK-16351 rtc: imx-sc: use SIP to set RTC time

commit  d2c777f7f84175b7c34215f9a86687066c5c9abe from
https://source.codeaurora.org/external/imx/linux-imx.git

For system controller RTC, as it belongs SC_R_SYSTEM,
and SC_R_SYSTEM is assigned in ARM-Trusted-Firmware,
so here needs to use SIP to trap into ATF to do set
time, or system controller firmware will return
error since linux kernel does NOT own this system
resource.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/rtc/rtc-imx-sc.c  |   27 ++++++++++-----------------
 include/soc/imx/fsl_sip.h |    3 +++
 2 files changed, 13 insertions(+), 17 deletions(-)

diff --git a/drivers/rtc/rtc-imx-sc.c b/drivers/rtc/rtc-imx-sc.c
index ce3add9..4fd2619 100644
--- a/drivers/rtc/rtc-imx-sc.c
+++ b/drivers/rtc/rtc-imx-sc.c
@@ -10,6 +10,7 @@
  * http://www.gnu.org/copyleft/gpl.html
  */
 
+#include <linux/arm-smccc.h>
 #include <linux/init.h>
 #include <linux/io.h>
 #include <linux/kernel.h>
@@ -18,6 +19,7 @@
 #include <linux/of_device.h>
 #include <linux/platform_device.h>
 #include <linux/rtc.h>
+#include <soc/imx/fsl_sip.h>
 #include <soc/imx8/sc/sci.h>
 
 sc_ipc_t timer_ipcHandle;
@@ -63,25 +65,16 @@ static int imx_sc_rtc_read_time(struct device *dev, struct rtc_time *tm)
 
 static int imx_sc_rtc_set_time(struct device *dev, struct rtc_time *tm)
 {
-	sc_err_t sciErr = SC_ERR_NONE;
+	struct arm_smccc_res res;
 
-	if (!timer_ipcHandle)
-		return -ENODEV;
+	/* pack 2 time parameters into 1 register, 16 bits for each */
+	arm_smccc_smc(FSL_SIP_SRTC, FSL_SIP_SRTC_SET_TIME,
+		((tm->tm_year + 1900) << 16) | (tm->tm_mon + 1),
+		(tm->tm_mday << 16) | tm->tm_hour,
+		(tm->tm_min << 16) | tm->tm_sec,
+		0, 0, 0, &res);
 
-	sciErr = sc_timer_set_rtc_time(timer_ipcHandle,
-		tm->tm_year + 1900,
-		tm->tm_mon + 1,
-		tm->tm_mday,
-		tm->tm_hour,
-		tm->tm_min,
-		tm->tm_sec);
-
-	if (sciErr) {
-		dev_err(dev, "failed to set time: %d\n", sciErr);
-		return -EINVAL;
-	}
-
-	return 0;
+	return res.a0;
 }
 
 static int imx_sc_rtc_read_alarm(struct device *dev, struct rtc_wkalrm *alrm)
diff --git a/include/soc/imx/fsl_sip.h b/include/soc/imx/fsl_sip.h
index 1d39f99..75459b1 100644
--- a/include/soc/imx/fsl_sip.h
+++ b/include/soc/imx/fsl_sip.h
@@ -20,6 +20,9 @@
 #define FSL_SIP_CPUFREQ			0xC2000001
 #define FSL_SIP_SET_CPUFREQ		0x00
 
+#define FSL_SIP_SRTC			0xC2000002
+#define FSL_SIP_SRTC_SET_TIME		0x00
+
 #define IMX8MQ_PD_MIPI		0
 #define IMX8MQ_PD_PCIE1		1
 #define IMX8MQ_PD_OTG1		2
-- 
1.7.9.5

